File: 120_Triangle/120_1.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>
#define MIN(a, b) ((a) < (b) ? (a) : (b))

int minimumTotal(int** triangle, int triangleSize, int* triangleColSize){
    for (int i = triangleSize - 2; i >= 0; --i) {
        for (int j = 0; j < triangleColSize[i]; ++j) {
            triangle[i][j] += MIN(triangle[i + 1][j], triangle[i + 1][j + 1]);
        }
    }
    return triangle[0][0];
}

int main(void) {
    int **triangle = (int **)malloc(sizeof(int *) * 4);
    triangle[0] = (int *)malloc(sizeof(int));
    triangle[0][0] = 2;
    triangle[1] = (int *)malloc(sizeof(int) * 2);
    triangle[1][0] = 3;  triangle[1][1] = 4;
    triangle[2] = (int *)malloc(sizeof(int) * 3);
    triangle[2][0] = 6; triangle[2][1] = 5; triangle[2][2] = 7;
    triangle[3] = (int *)malloc(sizeof(int) * 4);
    triangle[3][0] = 4; triangle[3][1] = 1; triangle[3][2] = 8; triangle[3][3] = 3;
    int colSizes[4] = {1, 2, 3, 4};
    printf("%d\n", minimumTotal(triangle, 4, colSizes));
}

File: 120_Triangle/120_1.java
import java.util.List;
import java.util.ArrayList;

class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int m = triangle.size();
        int n = triangle.get(m - 1).size();
        int[] dp = new int[n + 1];
        for (int i = m - 1; i >= 0; --i) {
            List<Integer> rows = triangle.get(i);
            for (int j = 0; j < rows.size(); ++j) {
                dp[j] = rows.get(j) + Math.min(dp[j], dp[j + 1]);
            }
        }
        return dp[0];
    }
}

public class leetcode {
    public static void main(String[] args) {
        List<Integer> t1 = new ArrayList<>();
        List<Integer> t2 = new ArrayList<>();
        List<Integer> t3 = new ArrayList<>();
        List<Integer> t4 = new ArrayList<>();
        t1.add(2);
        t2.add(3);
        t2.add(4);
        t3.add(6);
        t3.add(5);
        t3.add(7);
        t4.add(4);
        t4.add(1);
        t4.add(8);
        t4.add(3);
        List<List<Integer>> t = new ArrayList<>();
        t.add(t1);
        t.add(t2);
        t.add(t3);
        t.add(t4);
        Solution s = new Solution();
        System.out.println(s.minimumTotal(t));
    }
}